packages/dd-trace/src/profiling/util.js

```diff
@@ -8,6 +8,55 @@ function maybeRequire (id) {
   }
 }
 
+function eachSeries (collection, iteratee, callback) {
+  eachOfSeries(collection, (item, index, callback) => iteratee(item, callback), callback)
+}
+
+function eachOfSeries (collection, iteratee, callback = () => {}) {
+  const results = []
+  const next = index => {
+    if (collection[index]) {
+      iteratee(collection[index], index, (err, result) => {
+        if (err) return callback(err)
+
+        results.push(result)
+
+        next(index + 1)
+      })
+    } else {
+      callback(null, results)
+    }
+  }
+
+  next(0)
+}
+
+function parallel (tasks, callback) {
+  const results = new Array(tasks.length)
+
+  let counter = 0
+  let error = null
+
+  for (let i = 0; i < tasks.length; i++) {
+    const taskIndex = i
+
+    tasks[taskIndex]((err, result) => {
+      if (err && !error) {
+        error = err
+      }
+
```

Comment:
This function needs a shortcut for avoiding doing anything when an error is set. It also waits until all tasks are completed before calling the callback, even when there's an error, which is inefficient.

You can use https://npm.im/once (or your own implementation) to simplify this.

Reply:
> That works as designed, but maybe the design is wrong.

That may be the case here. It's adding to the results array when it's already known that the results array will never be used.

> I don't want to add once because I'm explicitly trying not to add more libraries to the tracer, otherwise I would just have used async.

Hence "or your own implementation". Here's one:

```js
function once(fn) {
  let called = false
  return (...args) => {
    if (called) return
    called = true
    fn(...args)
  }
}
```

Reply:
Fixed by keeping track of the failure state.